thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/detection/trigger/DataAvailabilityTaskScheduler.java

```diff
@@ -0,0 +1,216 @@
+          dataset2RefreshTimeMap.put(dataset, datasetConfig.getLastRefreshTime());
+        }
+      }
+    }
+  }
+
+  private Map<Long, TaskDTO> retrieveRunningDetection() {
+    List<TaskConstants.TaskStatus> statuses = new ArrayList<>();
+    statuses.add(TaskConstants.TaskStatus.WAITING);
+    statuses.add(TaskConstants.TaskStatus.RUNNING);
+    List<TaskDTO> tasks = taskDAO.findByStatusesAndTypeWithinDays(statuses, TaskConstants.TaskType.DETECTION,
+        (int) TimeUnit.MILLISECONDS.toDays(ThirdEyeUtils.DETECTION_TASK_MAX_LOOKBACK_WINDOW));
+    Map<Long, TaskDTO> res = new HashMap<>(tasks.size());
+    for (TaskDTO task : tasks) {
+      String[] parts = task.getJobName().split("_");
+      if (parts.length < 2) {
+        LOG.error("Invalid name found {} for task id {}", task.getJobName(), task.getId());
+        continue;
+      }
+      res.put(Long.parseLong(parts[1]), task);
+    }
+    return res;
+  }
+
+  private long createDetectionTask(DetectionConfigDTO detectionConfig) throws JsonProcessingException {
+    long detectionConfigId = detectionConfig.getId();
+    // Make sure start time is not out of DETECTION_TASK_MAX_LOOKBACK_WINDOW
+    long end = System.currentTimeMillis();
+    long start = Math.max(detectionConfig.getLastTimestamp(),
+        end - ThirdEyeUtils.DETECTION_TASK_MAX_LOOKBACK_WINDOW);
+    DetectionPipelineTaskInfo taskInfo = new DetectionPipelineTaskInfo(detectionConfigId, start, end);
+    String taskInfoJson = OBJECT_MAPPER.writeValueAsString(taskInfo);
+    TaskDTO taskDTO = new TaskDTO();
+    taskDTO.setTaskType(TaskConstants.TaskType.DETECTION);
+    taskDTO.setJobName(TaskConstants.TaskType.DETECTION.toString() + "_" + detectionConfigId);
+    taskDTO.setStatus(TaskConstants.TaskStatus.WAITING);
+    taskDTO.setTaskInfo(taskInfoJson);
+    long taskId = taskDAO.save(taskDTO);
+    LOG.info("Created detection pipeline task {} with taskId {}", taskDTO, taskId);
+    lastTaskCreateTimeMap.put(detectionConfigId, end);
+    return taskId;
+  }
+
+  private void loadLatestTaskCreateTime (DetectionConfigDTO detectionConfig) throws Exception {
+    long detectionConfigId = detectionConfig.getId();
+    List<TaskDTO> tasks = taskDAO.findByNameOrderByCreateTime(TaskConstants.TaskType.DETECTION.toString() +
+        "_" + detectionConfigId, 1,false);
+    if (tasks.size() == 0) {
+      lastTaskCreateTimeMap.put(detectionConfigId, 0L);
+    } else {
+      DetectionPipelineTaskInfo taskInfo = (DetectionPipelineTaskInfo) TaskInfoFactory.getTaskInfoFromTaskType(
+          TaskConstants.TaskType.DETECTION, tasks.get(0).getTaskInfo());
+      lastTaskCreateTimeMap.put(detectionConfigId, taskInfo.getEnd());
+    }
+  }
+
+  private boolean isAllDatasetUpdated(DetectionConfigDTO detectionConfig, Set<String> datasets,
+      Map<String, Long> dataset2RefreshTimeMap) {
+    long lastTimestamp = detectionConfig.getLastTimestamp();
+    return datasets.stream().map(d -> dataset2RefreshTimeMap.get(d) >= lastTimestamp)
```

Comment:
Maybe use `datasets.stream().allMatch(d -> dataset2RefreshTimeMap.get(d) >= lastTimestamp)` is cleaner

Reply:
Good point!